File: 프로그래머스/2/12905. 가장 큰 정사각형 찾기/가장 큰 정사각형 찾기.java
class Solution
{
    public int solution(int [][]board)
    {
        
        int answer = 0;
        
        int row = board.length;
        int col = board[0].length;
        
        if(row < 2) {
            for(int i = 0; i<col; i++){
                if(board[0][i] == 1){
                    return 1;
                }
            }
            return 0;
        }else if(col < 2){
            for(int i = 0; i<row; i++){
                if(board[i][0] == 1){
                    return 1;
                }
            }
            return 0;
        }
        
        for (int i = 1; i < row; i++){
            for(int j = 1; j<col; j++){
                if(board[i][j] != 0) {
                    board[i][j] = Math.min(board[i-1][j-1], Math.min(board[i][j-1], board[i-1][j])) + 1;
                        answer = Math.max(answer, board[i][j]);
                }
            }
        }

        
        

        return answer * answer;
    }
}

File: 프로그래머스/2/12905. 가장 큰 정사각형 찾기/가장 큰 정사각형 찾기.py
def solution(board):
    ans, row, col = 0, len(board), len(board[0])
    dp = [[0] * col for _ in range(row)]

    for i in range(row):
        for j in range(col):
            if i == 0 or j == 0:
                dp[i][j] = board[i][j]
            elif board[i][j] != 0:
                dp[i][j] = min(dp[i - 1][j - 1], dp[i - 1][j], dp[i][j - 1]) + 1

            ans = max(dp[i][j], ans)

    return ans * ans

print(solution([[0,1,1,1],[1,1,1,1],[1,1,1,1],[0,0,1,0]]))